### PR TITLE
[5.0] DataProtocol ContiguousCollection and new inline Data

### PR DESCRIPTION
**Pull request by itaiferber**
Cherry-pick of apple/swift#20225
(cherry picked from commit d0303543202236035ba85b74f2d538016462682e)

> This is an implementation of DataProtocol and a new smaller and leaner struct Data. The major change here is that we now promise that all Data's are contiguous. In order to enforce this requisite we added a new protocol to define contiguous collections (aptly named ContiguousCollection). This protocol is a draft of the work in progress proposed API for the standard library. Additionally it offers a direct byte accessor if safe on sequence to unify the sequence initialization to allow for fast paths.
> 
> Data itself now is represented in 16 bytes(on 64 bit platforms). This allows it to be stored in existentials without needing to heap allocate. Additionally during this refactor to shrink the representation we were able to make similar wins to the small string optimization so that we can store Data's on the stack when under a platform limit. This makes small Data structures never have to heap allocate. For example: Data() has zero heap allocations and Data([0xd, 0xa]) wont heap allocate for the storage of the data either since small representations can be stored completely on the stack efficiently. Larger data or slices however will still have a reference type storing the backing allocation as they did before. Our hope is that this will reduce the traffic on malloc which will reduce memory fragmentation and be considerably more efficient for small buffers.

### CODE REVIEW

**Comment by itaiferber** [comment]
Please test with following pull request:
apple/swift-lldb#1138

@swift-ci Please test

**Comment by itaiferber** [comment]
I've gone ahead and incorporated the changes from #21290 as well here.

**Comment by itaiferber** [comment]
Please test with following pull request:
apple/swift-lldb#1138

@swift-ci Please test

**Comment by itaiferber** [comment]
Please test with following pull request:
apple/swift-lldb#1138

@swift-ci Please test

File: stdlib/public/SDK/Foundation/Data.swift [hunk 1]
```diff
@@ -90,7 +78,7 @@ internal final class _DataStorage {
         }
     }
     
-    @usableFromInline
+    @inlinable
```
Comment (airspeedswift) [inline stdlib/public/SDK/Foundation/Data.swift]: Are you sure this needs to be inalienable? It's not a generic function.

File: stdlib/public/SDK/Foundation/Data.swift [hunk 2]
```diff
@@ -1244,35 +2055,71 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
...
-    public init<S: Sequence>(_ elements: S) where S.Iterator.Element == UInt8 {
-        let backing = _DataStorage(capacity: Swift.max(elements.underestimatedCount, 1))
-        var (iter, endIndex) = elements._copyContents(initializing: UnsafeMutableBufferPointer(start: backing._bytes?.bindMemory(to: UInt8.self, capacity: backing._capacity), count: backing._capacity))
-        backing._length = endIndex
-        while var element = iter.next() {
-            backing.append(&element, length: 1)
+    @inline(__always)
+    public init<S: Sequence>(_ elements: S) where S.Element == UInt8 {
+        // If the sequence is already contiguous, access the underlying raw memory directly.
+        if let contiguous = elements as? ContiguousBytes {
```
Comment (karwa) [inline stdlib/public/SDK/Foundation/Data.swift]: Just curious: why is this needed? Are there sequences of `UInt8`s which cannot implement `withContiguousStorageIfAvailable` but can conform to `ContiguousBytes`?

**Review by airspeedswift — APPROVED** [review]
Comments mostly are for future clean-up, though I would suggest inlinability could do with another pass to be certain it's justified everywhere. In some places the internals seem to be unnecessarily exposed in functions that don't seem critical to inline (apologies if experimentation has shown otherwise).

File: stdlib/public/SDK/Foundation/Data.swift [hunk 3]
```diff
@@ -844,221 +541,1334 @@ internal final class _DataStorage {
...
-    
-#if !DEPLOYMENT_RUNTIME_SWIFT
-    @objc override
-    func _isCompact() -> Bool {
-        return true
-    }
-#endif
-    
-#if DEPLOYMENT_RUNTIME_SWIFT
-
```
[the remark this answers is not in the document]
Comment (airspeedswift) [reply stdlib/public/SDK/Foundation/Data.swift]: weird, it looks ok on the code tab

File: stdlib/public/SDK/Foundation/DataProtocol.swift [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+//===--- DataProtocol Extensions ------------------------------------------===//
+
+extension DataProtocol {
+    public func firstRange<D: DataProtocol>(of data: D) -> Range<Index>? {
+        return self.firstRange(of: data, in: self.startIndex ..< self.endIndex)
+    }
+
+    public func lastRange<D: DataProtocol>(of data: D) -> Range<Index>? {
+        return self.firstRange(of: data, in: self.startIndex ..< self.endIndex)
```
Comment (karwa) [inline stdlib/public/SDK/Foundation/DataProtocol.swift]: Should be `lastRange`

File: stdlib/public/SDK/Foundation/Data.swift [hunk 5]
```diff
@@ -844,221 +541,1334 @@ internal final class _DataStorage {
...
+        @inlinable
+        mutating func resetBytes(in range: Range<Index>) {
+            assert(range.lowerBound <= MemoryLayout<Buffer>.size)
+            assert(range.upperBound <= MemoryLayout<Buffer>.size)
+            precondition(range.lowerBound <= length, "index \(range.lowerBound) is out of bounds of 0..<\(length)")
+            if count < range.upperBound {
+                count = range.upperBound
+            }
+            Swift.withUnsafeMutableBytes(of: &bytes) { rawBuffer in
+                bzero(rawBuffer.baseAddress?.advanced(by: range.lowerBound), range.upperBound - range.lowerBound)
```
Comment (karwa) [inline stdlib/public/SDK/Foundation/Data.swift]: `bzero` is deprecated and doesn't exist on some platforms (e.g. windows). Use `memset(3)` instead.

http://pubs.opengroup.org/onlinepubs/009696699/functions/bzero.html

**Comment by karwa** [comment]
This is an interesting design and the performance improvements are welcome, but I'd nonetheless like to register my severe disappointment that this major change to the Swift API never went through swift-evolution.

File: stdlib/public/SDK/Foundation/Data.swift (hunk 2 above)
Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Data.swift]: Yes: any sequence of `UInt8`'s which does not represent _typed_ memory cannot implement `withContiguousStorageIfAvailable` without handing over a copy, which defeats what we need to do here. `Data` and `UnsafeRawBufferPointer` are both examples of such sequences — both have an element type of `UInt8`, but represent raw memory, and thus cannot implement `withContiguousStorageIfAvailable`.

File: stdlib/public/SDK/Foundation/Data.swift [hunk 6]
```diff
@@ -844,221 +541,1334 @@ internal final class _DataStorage {
...
+        var range: Range<Int>
+
+        @inlinable
+        var lowerBound: Int { return range.lowerBound }
+
+        @inlinable
+        var upperBound: Int { return range.upperBound }
+
+        @inlinable
+        var count: Int { return range.upperBound - range.lowerBound }
```
[the remark this answers is not in the document]
Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Data.swift]: Re: this and `range.isEmpty` — we found during testing that because `range` does not implement `count` itself (but instead gets the implementation from a default implementation on `Collection`), calculating this ourselves was orders of magnitude faster than calling `range.count` (dispatching through to the protocol implementation added a lot of overhead), and this was making a big difference in our unit tests.

It does appear that `Range` implements `isEmpty` directly, so we should be able to fix `.isEmpty` above.

File: stdlib/public/SDK/Foundation/DataProtocol.swift (hunk 4 above)
Comment (itaiferber) [reply stdlib/public/SDK/Foundation/DataProtocol.swift]: Good catch!

File: stdlib/public/SDK/Foundation/Data.swift (hunk 5 above)
Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Data.swift]: This implementation is only for Darwin platforms so `bzero` is fine, but I'll make this change so things are easier to copy over to swift-corelibs-foundation wholesale.

**Comment by itaiferber** [comment]
@karwa The changes here are Foundation API, and the corelibs process is separate from swift-evolution. If we decide at any point to integrate this work into the standard library directly, we'll certainly go through swift-evolution, just like everyone else.

**Comment by itaiferber** [comment]
@airspeedswift Thanks for the review! I'm going to make some of the functional changes here that actually affect the implementation — as you mention, we can do the inlinability fixes in another pass very soon. We'd like to get the implementation landed first for testing and we have a little wiggle room to re-evaluate inlinability.

In general, we are aggressively inlining to give the compiler as much room as possible to make performance optimizations, which we've seen are largely necessary to make this work compelling. We can evaluate on a case-by-case basis if we see this having negative effects.

File: stdlib/public/SDK/Foundation/Data.swift [hunk 7]
```diff
@@ -844,221 +541,1334 @@ internal final class _DataStorage {
...
+#else
+    @objc(_providesConcreteBacking)
+    func _providesConcreteBacking() -> Bool {
+        return true
+    }
+#endif
+}
+
+@_fixed_layout
+public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessCollection, MutableCollection, RangeReplaceableCollection, MutableDataProtocol, ContiguousBytes {
```
[the remark this answers is not in the document]
Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Data.swift]: I agree — I didn't want to take on the task right now, but we should do this in the future.

File: stdlib/public/SDK/Foundation/Data.swift [hunk 8]
```diff
@@ -844,221 +541,1334 @@ internal final class _DataStorage {
...
+
+        @inlinable
+        var capacity: Int {
+            return MemoryLayout<Buffer>.size
+        }
+
+        @inlinable
+        var count: Int {
+            get {
+                return numericCast(length)
```
[the remark this answers is not in the document]
Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Data.swift]: I'll change this too

**Comment by airspeedswift** [comment]
> In general, we are aggressively inlining to give the compiler as much room as possible to make performance optimizations

A word of warning – this can backfire. Marking things inlinable sometimes slows code down unexpectedly too because it limits what assumptions the compiler can make about internal uses of the inlinable function. It also bloats user binary size.

File: stdlib/public/SDK/Foundation/Data.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (airspeedswift) [reply stdlib/public/SDK/Foundation/Data.swift]: Can you file a bug for the optimizer for that? We should try and ensure they're equivalent if possible. It might also be something that inlining too much is causing as the inliner might be giving up on a huge function that needs a specialized `count` the middle.

**Comment by itaiferber** [comment]
Please test with following pull request:
apple/swift-lldb#1138

@swift-ci Please test

**Comment by airspeedswift** [comment]
I'm OK to merge this. Can you add some tests for the parts that had bugs that didn't get test failures in a follow-up PR?

File: stdlib/public/SDK/Foundation/Data.swift [hunk 9]
```diff
@@ -1244,35 +2049,71 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
...
+                let capacity = buffer.pointee.capacity
+                let base = buffer.pointee.memory.bindMemory(to: UInt8.self, capacity: capacity)
+                var (iter, endIndex) = elements._copyContents(initializing: UnsafeMutableBufferPointer(start: base, count: capacity))
+
+                // Copy the contents of buffer...
+                _representation = _Representation(UnsafeRawBufferPointer(start: base, count: endIndex))
+
+                // ... and append the rest byte-wise.
+                while let element = iter.next() {
+                    Swift.withUnsafeBytes(of: element) {
```
Comment (palimondo) [inline stdlib/public/SDK/Foundation/Data.swift]: This looks like a terrible performance pitfall to me... No benchmarks currently cover this, but in my local test, this is 32x slower than hitting the happy `_copyContents` path.

Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Data.swift]: Sorry for the delay in responding — just got back from vacation.
Just to clarify, do you mean by this comment that we should slurp all of the elements out of the iterator into an array, then append to `_representation`, or just that hitting this code path is slow? (Because the latter is somewhat expected — we've got a sequence here which is not `ContiguousBytes`, doesn't have an implementation of `withContiguousStorageIfAvailable`, _and_ doesn't implement the `_copyContents` optimization, AKA, the worst-case scenario.)

Comment (palimondo) [reply stdlib/public/SDK/Foundation/Data.swift]: The first. I don't think such slowdown is acceptable. Here's benchmark I got:

````swift
…
  BenchmarkInfo(name: "Data.append.Sequence.ExactCount",
    runFunction: { append($0*100,
      sequence: repeatElement(UInt8(0xA0), count: 809), to: medium) }, tags: d),
  BenchmarkInfo(name: "Data.append.Sequence.UnderestimatedCount",
    runFunction: { append($0*100,
      sequence: repeatElementSeq(809), to: medium) }, tags: d),

…

let repeatElementSeq = { count in
  return sequence(state: count) { (i: inout Int) -> UInt8? in
    defer { i = i &- 1 }; return i > 0 ? UInt8(0xA0) : nil
  }
}

@inline(never)
func append<S: Sequence>(_ N: Int, sequence: S, to data: Data)
  where S.Element == UInt8 {
  for _ in 1...N {
    var copy = data
    copy.append(contentsOf: sequence)
  }
}
````

Comment (palimondo) [reply stdlib/public/SDK/Foundation/Data.swift]: ...and just to clarify, we probably don't need to slurp **all of the elements** out of the iterator in one go, just amortize the costs of going over the `UnsafeBufferPointer` by slurping more elements in batches periodically append to the `_representation`. We could try reusing `InlineData`, or test some moderately sized `ContiguousArray`…

Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Data.swift]: @palimondo Sure, this is a reasonable change we can and should make. Are you interested in taking this on?

(We should make a similar change in `append<S : Sequence>(_: S)` as well.)

Comment (palimondo) [reply stdlib/public/SDK/Foundation/Data.swift]: Not at the moment. I want to first finish my cleanup pass of the Swift Benchmark Suite, applying the `legacyFactor`.

Also, when I have played with this locally (I was `println` debugging `Data` implementation to see which branch get's exercised), I had trouble getting the build to work properly. Maybe it was just that most of the impl was inlined and the DataBenchmarks were't getting rebuilt? It was a mess.

So I cannot promise I could get back to this sooner than in two weeks...

Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Data.swift]: @palimondo No worries, I'll take care of this, possibly in #21754
